src/constraint_system/composer.rs

```diff
@@ -128,13 +128,23 @@ impl TurboComposer {
     }
 
     /// Evaluate the runtime value of a witness
-    #[allow(dead_code)]
-    pub(crate) fn evaluate_witness(&self, witness: &Witness) -> &BlsScalar {
-        &self.witnesses[&witness]
+    ///
+    /// # Safety
+    ///
+    /// Using the allocated scalar of a witness inside a gadget could produce
+    /// an unsound circuit.
+    pub unsafe fn evaluate_witness(&self, witness: &Witness) -> &BlsScalar {
+        &self.witnesses[witness]
+    }
+
+    /// Add a new witness with the same scalar of the given one
+    pub fn duplicate_witness(&mut self, witness: &Witness) -> Witness {
```

Comment:
Please, elaborate on why this functionality is needed.

This is not particularly secure because the new witness isn't constrained to the previous one. We often want to avoid at all costs creating new witnesses, and we can do so only if we have a clear constraint that will make that operation safe.

Reply:
In zelbet we need to append a new witness based on the value of another witness. One solution is indeed having:
```rust
let w1 = composer.append_witness(unsafe { composer.evaluate_witness(w) });
```

But since we don't actually care about leaking the allocated scalar – we do not perform operation with that, we just care about the witness – having a function that avoid to expose the scalar sounded better.

but we don't actually care what is the allocated scalar.

Reply:
Since `duplicate_witness` is also unsafe (e.g. the previous constraints are not applied to the new witness and the circuit will still be valid), the snippet suggestion is a better approach